### ISSUE TITLE
"operation's operand is unlinked" error in iree-stream-schedule-concurrency

### ISSUE THREAD

**Issue by phoenix-meadowlark** [iree-org/iree#13648]
### What happened?

`iree-compile` throws the following error during `iree-stream-schedule-concurrency` when compiling the training step of a GPT-3-like LLM to the LLVM-CPU backend:

```
<unknown>:0: error: operation's operand is unlinked
<unknown>:0: note: see current operation: "func.return"(%674#37, %517#31, %517#30, %517#29, %517#28, %517#27, %517#26, %517#25, %517#24, %517#23, %517#22, %517#21, %517#20, %517#19, %517#18, %517#17, %517#16, %674#46, %674#47, %674#48, %674#1, %674#2, %674#3, %674#4, %674#5, %674#6, %674#7, %674#8, %674#9, %674#10, %674#49, %674#11, %674#12, %674#13, %674#14, %674#15, %674#16, %674#17, %674#18, %674#19, %674#20, %674#21, %674#22, %674#23, %674#24, %674#25, %674#26, %674#27, %674#28, %674#29, %674#30, %674#31, %674#32, %674#33, %674#34, %674#35, %674#36, %676, %757#0, %757#1, %674#40, %758, %674#42, %755, %674#41, %756, %674#43, %754, %674#39, %759, %677, %675, %674#50, %674#45, %517#32, %674#44, %674#0, <<UNKNOWN SSA VALUE>>, %674#51, %753, %752, %751, %750, %749, %748, %747, %746, %745, %744, %743, %742, %741, %740, %739, %738, %737, %736, %735, %734, %733, %732, %731, %730, %729, %728, %727, %726, %725, %724, %723, %722, %721, %720, %719, %718, %717, %716, %715, %714, %713, %712, %711, %710, %709, %708, %707, %706, %705, %704, %703, %702, %701, %700, %699, %698, %697, %696, %695, %694, %693, %692, %691, %690, %689, %688, %687, %686, %685, %684, %683, %682, %681, %680, %679, %678, %674#52, %674#53, %674#54, %674#55, %674#56, %674#57, %674#58, %674#59, %674#60, %674#61, %674#62, %674#63, %674#64, %674#65, %674#66, %674#38) : (!stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>, !stream.resource<external>) -> ()
```

The IR produced by `--mlir-print-ir-before-all` can be found here: [`before_schedule_concurrency.mlir`](https://gist.github.com/phoenix-meadowlark/d0193e93c8d7b60082a431a57c89ed8b)

### Steps to reproduce your issue

Download [`gpt_train_step.mlir`](https://gist.github.com/phoenix-meadowlark/ce1ca6ed0a60a1a08a4c6359ab91cf32) and execute the following command:

```shell
iree-compile \
  --iree-hal-target-backends=llvm-cpu \
  --iree-input-type=mhlo \
  /tmp/gpt_train_step.mlir \
  -o /tmp/gpt_train_step.vmfb
```

### What component(s) does this issue relate to?

Compiler

### Version information

Installed via PIP:

```
iree-compile --version
IREE (https://openxla.github.io/iree):
  IREE compiler version 20230515.520 @ 26d9eb8209077fd56befe6dbc052d463187a9c4e
  LLVM version 17.0.0git
  Optimized build
```

**Comment by phoenix-meadowlark** [comment]
Confirmed that this also affects the CUDA backend, but that seems expected.

**Comment by benvanik** [comment]
Can you try at head? I was able to successfully compile the input through the stream dialect (`--compile-to=stream`) and it's possible one of my changes yesterday fixed this (#13625 maybe).

**Comment by phoenix-meadowlark** [comment]
Looks like that fixed it!
